novawallet/Common/URLHandling/Parsing/OpenDAppUrlParsingService.swift

```diff
@@ -0,0 +1,61 @@
+import Foundation
+import RobinHood
+
+final class OpenDAppUrlParsingService: OpenScreenUrlParsingServiceProtocol {
+    private let dAppProvider: AnySingleValueProvider<DAppList>
+
+    enum Key {
+        static let url = "url"
+    }
+
+    convenience init() {
+        let dAppsUrl = ApplicationConfig.shared.dAppsListURL
+        let dAppProvider: AnySingleValueProvider<DAppList> = JsonDataProviderFactory.shared.getJson(
+            for: dAppsUrl
+        )
+        self.init(dAppProvider: dAppProvider)
+    }
+
+    init(dAppProvider: AnySingleValueProvider<DAppList>) {
+        self.dAppProvider = dAppProvider
+    }
+
+    func parse(url: URL) -> Result<UrlHandlingScreen, DeeplinkParseError> {
+        guard let urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false),
+              let query = urlComponents.queryItems else {
+            return .failure(.openGovScreen(.emptyQueryParameters))
+        }
+
+        let dAppUrl = query.first(where: {
+            $0.name
+                .trimmingCharacters(in: .whitespacesAndNewlines)
+                .caseInsensitiveCompare(Key.url) == .orderedSame
+        })?.value.map { URL(string: $0) }
+
+        guard let dAppUrl = dAppUrl else {
+            return .failure(.openDAppScreen(.invalidURL))
+        }
+
+        let semaphore = DispatchSemaphore(value: 0)
+
+        var result: Result<UrlHandlingScreen, DeeplinkParseError>?
```

Comment:
I think it is better to pass url to the DAppListPresenter and then wait when metadata comes from the interactor. Similarly what you did with governance. The main reason is that we are blocking main thread by waiting the database. Also we might run into the situation when the database is not up to date.

Reply:
Let's fix the issue in gov PR